### ISSUE TITLE
TorchBench V2 nightly run detects a 7.8% regression on nvidia_deeprecommender model

### ISSUE THREAD

**Issue by xuzhao9** [pytorch/pytorch#73328]
### 🐛 Describe the bug

On 20220223, the TorchBench test test_train[nvidia_deeprecommender-cuda-eager] regresses 7.89634% because of the commit https://github.com/pytorch/pytorch/commit/3aecce7015.

The perf delta has been validated by PR: https://github.com/pytorch/pytorch/pull/73306. After reverting the commit, the nvidia_deeprecommender tests show speedup. We observe the latency difference below (lower the better):

```
Benchmark                                       Run 6688487         Run 41bdfa78      Delta
----------------------------------------------  ----------------  -----------------  ------------
test_train[nvidia_deeprecommender-cpu-eager]       0.455588              0.460329   +1.030094%
test_eval[nvidia_deeprecommender-cpu-eager]        0.0818041             0.0791861  -3.306099%
test_train[nvidia_deeprecommender-cuda-eager]     0.131359                0.121776   -7.868537%*
test_eval[nvidia_deeprecommender-cuda-eager]      0.0369564            0.0289373  -27.711980%*
```

Commit 6688487 contains the commit 3aecce7015, and commit 41bdfa78 reverts 3aecce7015. After reverting the commit, the latency of test `nvidia_deeprecommender` decreases (show there is speedup).
The CPU test delta is below the noise threshold `7%`, but the CUDA tests show performance signals.

### Versions

PyTorch 1.12.0.dev20220223+cu113

cc @VitalyFedyunin @ngimel

**Comment by ngimel** [comment]
@xuzhao9, what hardware and datatype are used for the benchmarking?
cc @jspark1105

**Comment by xuzhao9** [comment]
> @xuzhao9, what hardware and datatype are used for the benchmarking? cc @jspark1105

I am using Nvidia T4 16GB, datatype fp32.

**Comment by jspark1105** [comment]
@xuzhao9 Thanks for reporting this issue. Do you happen to know what GEMM shapes this benchmark is running (m, n, k, and transpose settings for A/B), and measurements on other HWs like V100/A100?

**Comment by ngimel** [comment]
cc @xwang233, what's up with cublasLt?

**Comment by xwang233** [comment]
Can someone point me an end-to-end reproduction script for the regressed model? Detailed GEMM shapes would be more helpful. Thanks!

**Comment by xuzhao9** [comment]
> Can someone point me an end-to-end reproduction script for the regressed model? Detailed GEMM shapes would be more helpful. Thanks!

Here is the script to reproduce, assuming you are using Python 3.8, CUDA 11.3:

```
pip install --pre torch torchtext torchvision \
          -f https://download.pytorch.org/whl/nightly/cu113/torch_nightly.html
git clone git@github.com:pytorch/benchmark.git; pushd benchmark
python install.py nvidia_deeprecommender
python run.py nvidia_deeprecommender -d cuda -t train
```

Sorry but Python 3.9+ won't work for now.

**Comment by xuzhao9** [comment]
@jspark1105 @xwang233 I attached the CUDA profile here. Please let me know if it contains the GEMM shapes you need.
[profile.tar.gz](https://github.com/pytorch/pytorch/files/8128614/profile.tar.gz)

**Comment by ngimel** [comment]
@xwang233 how should we adjust the criteria to call fused bias add? It would be easy to use fusion only for half datatype e.g. because that's the primary perf-sensitive usecase, and completely disable it for fp32, but maybe there's a more targeted fix? Or maybe, if cublas doesn't do fusion for fp32, it just makes sense to disable cublasLt for fp32?

**Comment by xuzhao9** [comment]
The performance is recovered after https://github.com/pytorch/pytorch/pull/73390

**Comment by xwang233** [comment]
Hi @jspark1105 , do you know the matrix shape M, N, K of the 197951x512 GEMM operation?

**Comment by jspark1105** [comment]
@xwang233 it's for m=256, n=512, k=197951

**Comment by xwang233** [comment]
The cublasLt workspace is set to zero here. https://github.com/pytorch/pytorch/blob/3f78ec3bd422fa69a4ade6eee700afc032d4bd23/aten/src/ATen/cuda/CUDABlas.cpp#L691

From cublas team:

Generally, it's recommended to use 4 MB of workspace. For this specific shape (256, 512, 197951), even a small workspace as 16 KB could help.

cuBLAS has its internal workspace pool and that's why it doesn't have this issue. cuBLASLt is "stateless", and user needs to provide the workspace explicitly.

**Comment by jspark1105** [comment]
@xwang233 thanks for the information and I can confirm that by setting workspaceSize = 4 * 1024 * 1024, I can also recover the perf. Sorry I misinterpreted that cublasLtMatmulPreferenceSetAttribute sets workspaceSize and it doesn't matter what value workspaceSize is set to before.
@ngimel would you recommend to keep workspace as a global state (4MB is not bad) or create for each call like the current code would do (https://github.com/pytorch/pytorch/blob/3f78ec3bd422fa69a4ade6eee700afc032d4bd23/aten/src/ATen/cuda/CUDABlas.cpp#L698-L700) relying on the caching allocator?

**Comment by ngimel** [comment]
Let's create per-op, to be consistent with other functions, also, we can probably reduce it to 1MB.

**Comment by xuzhao9** [comment]
Closed as we are disabling cublasLT in 11.3.

**Comment by hdhai** [comment]
> Let's create per-op, to be consistent with other functions, also, we can probably reduce it to 1MB.

I have a question:  why don't you use real workspaceSize  returned by cublasLtMatmulAlgoGetHeuristic to allocate, but use the default 1M size ?  Default size is global setting, but we know in different cases，real workspaceSize of cublas is also different.   Looking forward to your reply.
